### ISSUE TITLE
Question regarding build_targets

### ISSUE THREAD

**Issue by ydixon** [eriklindernoren/PyTorch-YOLOv3#53]
I found that `conf_mask[b, anch_ious > ignore_thres] = 0` will overwrite any previous ground truth target that has `conf_mask` set to 1.

        for t in range(target.shape[1]):
            if target[b, t].sum() == 0:
                continue
            .......
            # Calculate iou between gt and anchor shapes
            anch_ious = bbox_iou(gt_box, anchor_shapes)
            # Where the overlap is larger than threshold set mask to zero (ignore)
            conf_mask[b, anch_ious > ignore_thres] = 0    <---------------------
            # Find the best matching anchor box
            best_n = np.argmax(anch_ious)
            ...
            # Masks
            mask[b, best_n, gj, gi] = 1
            conf_mask[b, best_n, gj, gi] = 1    <-------------------

Let's say: 
```
Iter1: anch_ious = [0.567 0.305 0.43], anch_ious>ignore_thres=0,best_n=0, conf_mask[b, 0, gj, gi]= 1
Iter2: anch_ious = [0.667 0.045 0.33], anch_ious>ignore_thres=0,best_n=0, conf_mask[b, 0, gj, gi]= 1
Iter3: anch_ious = [0.734 0.025 0.22], anch_ious>ignore_thres=0,best_n=0, conf_mask[b, 0, gj, gi]= 1
```

So every iteration is erasing the previous iteration's ground truth target's `conf_mask` if the target is selecting the same anchor. For the example above, only `Iter3: conf_mask[b, 0, gj, gi]= 1`is kept. It seems like more than half of the ground truth target's are usually ignored during training. Is this intended behavior and why would it work for training?

**Comment by Aslizy** [comment]
It is not erasing, more like for one truth bb should not correspond to two anchors boxes. So  conf_mask[b, anch_ious > ignore_thres, gj, gi] = 0 initialize those possible anchor boxes to 0, and then in the later step conf_mask[b, best_n, gj, gi] = 1 assign the best iou anchor box to that truth bb. 

In your example, I guess in each iteration, it is computing the anch_ious of each truth bb (for t in range(target.shape[1]). So you expect the gj and gi are different for those bbs. (under the assumption that it is not the case of multiple-label assignment). So what conf_mask[b, anch_ious > ignore_thres, gj, gi] = 0 does is not erasing.

**Comment by ydixon** [comment]
I think either way the best iou anchor box is only assigned to one truth bb. Because if there's repeating assignment on the anchor box, the conf_mask and tcls would only be set to 1 again. 

However, for the case above, it will wipe all `anch_ious > ignore_thres` even the ones that have another truth bb assigned.

**Comment by eriklindernoren** [comment]
An anchor for a given grid cell can only be assigned to one object, so that would be fine in this case, given that the specific anchor is the best match for these objects.
